### ISSUE TITLE
TestContext.CurrentTest exposes too much internal info

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#1578]
The following problems should be fixed... and the code examined for any others...
1. The Properties property exposes IPropertyBag, which is intended as an extensibility interface, not one for users. Among other things, it allows setting properties, which is not a good idea and potentially confusing since they would not be persisted. The internal PropertyBag should be wrapped by an adapter object, just as we wrap the test and the result.
2. The Properties indexer returns an IList. We should determine if this can be made type-safe. If it cannot be done at the present time (IIRC, we have both string and int properties at a minimum) then this should be spun out as a separate future issue.
3. The IList returned by the indexer gives too much access. An IEnumerable would be better. Alternatively, another layer of adapter with enumeration and Count available would do the job.

Hard to know how to prioritize this. I'm marking it Low for now, but if we see other issues arising that relate to this problem we should increase it.

**Comment by jnm2** [comment]
re: https://github.com/nunit/nunit/issues/2665#issuecomment-358545450

> While the question warrants some discussion, I don't think it needs a log. It's an error and the breaking change will be a failure to compile. That's a good thing since changing the property bag doesn't currently have any effect. For example, if you set the MaxTime property, you don't suddenly get MaxTime behavior because that was all decided before the test ever ran. IOW, it's a bug to fix.

I'm not concerned for the sake of expecting MaxTime or other behaviors to be retroactively triggered. I'm concerned that someone will have been "fixing" values or adding their own custom names to the bag. We've seen this with test parameters if I remember right. Is that clearly an error?

**Comment by CharliePoole** [comment]
See correction to my comment above: s/log/lot/

I'm not sure what you mean WRT test parameters. For properties, the current code allows access right into the property bag associated with the test object. It's possible that adding or changing a property would be reflected in the XML output. It's equally likely in my mind that you could crash the test run by changing the properties of a test in the middle of its execution. It would take a lot of experimentation to figure out the limits where this could be done safely and I don't think it's worth the effort because exposing the properties to change was an __error__ in the first place. We've discussed this in several issues in the past, but I'm damned if I can find one of them now that I need it! We also discussed creation of __result properties__, which could be modified or created by the test but have never acted on it.

Bottom line, the possibility of changing properties was an error I introduced in the first place by simply pointing to the real property bag rather than copying it or providing a read-only interface. Even though it's possible that some user has followed up on my error by changing the properties, I feel no compunction in breaking their code because there is no documentation that suggests they should feel free to do that. I know, we have made exceptions to this in the past, particularly when the bad code came to be relied on by a third-party application with lots of users. I don't think this is such a case, however.

This is about the amount of discussion I thought we needed. Is it enough for everyone else?

We've had lots of discussion on t

**Comment by jnm2** [comment]
The things to discuss in my mind are: 

- What about the possibility of keeping it IPropertyBag but providing an internal adapter implementation which throws for mutating methods? That would keep the code compiling and fail only the misbehaving tests with a diagnostic message. This feels less clean to me than breaking.

- If we don't mind breaking source code, should we add obsoleted methods on the new type to match the mutating methods on IPropertyBag with `[Obsolete("Diagnostic message, error: true)]` rather than letting people wonder what exactly changed?

**Comment by CharliePoole** [comment]
If Properties returns a type that includes the methods we want used, then nothing breaks except code that uses methods we __don't__ want used. It shouldn't matter in this case whether the interface is `IPropertyBag` or `IReadOnlyPropertyBag` or if there is no interface at all...

```C#
TestContext.CurrentContext.Test.Properties.Get("SomeProp"); // keeps working
TestContext.CurrentContext.Test.Properties.Set("SomeProp"); // won't compile
```

Alternatives to simply leaving out `Set` are
1. Make it obsolete. This is a misnomer, since it's not something we __intentionally__ ever supported, but it is a way to guarantee a clear error message.
2. Throwing an exception at runtime. This also gives us control over the error message but the user doesn't see it as soon.

I could live with any of the alternatives. My order of preference is (1) Just leave the method out (2) Use Obsolete (3) Throw. The reason I like the first choice best is that we are then done with this and don't have to revisit it again and I'm betting that not many users are affected.

I'd feel differently if I hadn't been involved in so many discussions about how setting the properties will not work or (especially) if our documentation suggested you could effectively change the properties.

**Comment by jnm2** [comment]
I didn't clarify this in this thread, but unless we go with keeping the type `IPropertyBag` and having the mutating methods throw at runtime, code that has always used this API **properly** will break:

```cs
var foo = ExtractSpecialProperty(TestContext.CurrentContext.Test.Properties);

// ...

Foo ExtractSpecialProperty(IPropertyBag properties)
{
    // ...
}
```

**Comment by jnm2** [comment]
That's not true actually. Another option to avoid breaking that legitimate souce code is make the new type a class and add an implicit conversion which adapts it to a readonly IPropertyBag on demand.

**Comment by jnm2** [comment]
There are scenarios where legitimate user code is broken even with the implicit cast, but so far none I've thought of are credible to me.

**Comment by jnm2** [comment]
Ignore the part about the implicit cast. User-defined conversions to an interface type are not supported in C#.

**Comment by CharliePoole** [comment]
Let me answer your objection in more general terms. In my view it is __always__ possible to create an example with the aim of breaking something. It is, therefore, always possible for any public change to be called a "breaking change" in the loosest sense. But we call such examples "contrived" for a reason. Most users are not __trying__ to break things and I want to avoid "breaking changes" in a much more restricted meaning of the term: changes that break the code of users who are using the feature as it was designed to be used.

So...

It's absolutely true that __if__ somebody has been passing the value of the `Properties` property to some method that takes an `IPropertyBag` as an argument, then the code will not compile. To that I will say two things:

1. This is not the normal use of the `Properties` property. I seriously doubt that it will affect many users. If you have an example to indicate otherwise, put it out there.

2. For the users it affects, breaking their compile is the best thing we can do for them. It shows the error immediately and in an unambiguous way.

Sorry if I got a bit impatient with your eamples. I was starting from the point of view that it's __obvious__ that the change will break some code. That's the difference between a change that breaks things and a "BreakingChange" as we typically use the term... a change in how the user are __expected__ to use the feature and which we must try to avoid at all costs.

**Comment by jnm2** [comment]
And that's exactly what I'm hoping for discussion on. How likely do we think it is that legitimate user code is doing anything other than simply dotting through the `ITest.Properties` property? Is this something you would see yourself writing?

```cs
IPropertyBag properties = TestContext.CurrentContext.Test.Properties;
anObjectWithALongName.AMethodWithAReallyLongName(properties, "Parameters that take up a lot of space", 42, foo);
```

**Comment by OmicronPersei** [comment]
Well, what can you get functionally out of incorrectly using `TestContext.CurrentContext.Test.Properties`?  Assuming that's disabled, would there be some viable workaround for users affected?

(Honestly, I think there's only so much we can get out of speculating what the NUnit community at large does until we actually go out and ask them in some manner, or wait for complaints if we implement some level of a "breaking" change.)

**Comment by jnm2** [comment]
> Well, what can you get functionally out of incorrectly using TestContext.CurrentContext.Test.Properties? 

Nothing that I know of, besides affecting your own test code if it reads properties later on.

> or wait for complaints if we implement some level of a "breaking" change

Fyi we have had some of this in the past. Choosing this strategy is fine so long as we choose it deliberately.

Okay, I don't want to hold this up. This is the only type of break we'll probably see and correcting it should be pretty easy to figure out: `IPropertyBag properties = TestContext.CurrentContext.Test.Properties;` You have Charlie's and my support.

**Comment by ChrisMaddock** [comment]
I'm happy with that plan.

**Comment by OmicronPersei** [comment]
> The Properties indexer returns an IList. We should determine if this can be made type-safe.

What does type-safe mean in this context?  Sorry, probably a newb question on my part.  :laughing:

**Comment by CharliePoole** [comment]
I've long thought we should only expose string properties, since they are embedded in an XML string anyway. In that case, we could return `IList<string>`  or `IEnumerable<string>`.

Internally, we have used other types as properties, particularly ints and enums. The question is whether the user needs to get them as ints, etc. or if a string representation is sufficient here.

Since users cannot add properties, the Types we use are the only ones that matter.

**Comment by jnm2** [comment]
Users can add properties by writing attributes though. It seems more likely to me that users would read properties of their own attributes than NUnit's internal properties (which really need a case-by-case API to read).

**Comment by CharliePoole** [comment]
You're right. I guess we're stuck with either IList or IList<object>. WRT internal properties, they were all designed to be used by NUnit itself, although visible to users. They are given to runners as strings (in the XML). In cases where I felt we needed an API of some kind, I didn't use a property. That's not to say there can't be errors somewhere in the design, but that was the intention anyway.

**Comment by jnm2** [comment]
(Edited to see what you had written because the markdown parser thought \<object> was an HTML element.)

Have we used readonly ILists before? Declaring our own `IReadOnlyList<>` for net40 and under would be cool...

**Comment by OmicronPersei** [comment]
>Users can add properties by writing attributes though.

I thought the intent here was to disable any functionality like this?

**Comment by jnm2** [comment]
@OmicronPersei The intent is for the properties to be readonly once the test is executing. When attributes add properties, they don't access TestContext. The control is inverted and the framework reads the attributes' Properties collections and copies them into the TestContext prior to running.

**Comment by MaceWindu** [comment]
FYI, this change broke NUnit.ApplicationDomain extension:
https://bitbucket.org/zastrowm/nunit.applicationdomain/src/1ad5ef9164244109c62c71db6691ef501500dccc/src/NUnit.ApplicationDomain/NUnit.Framework/AppDomainRunner.cs?at=master&fileviewer=file-view-default

**Comment by jnm2** [comment]
@nunit/framework-team We broke an extension that relies on storing a lazily-initialized instance in TestContext.CurrentContext.Test.Properties. This is a binary-breaking change (TypeLoadException for NUnit.ApplicationDomain) as well as a source-breaking change (NUnit.ApplicationDomain will not compile against 3.10).

- Is there an alternative approach NUnit.ApplicationDomain should be using?

- Is there anything we should do to make life easier for users who want to use both NUnit 3.10 and current or past versions of NUnit.ApplicationDomain?

**Comment by jnm2** [comment]
This tells me I should be more annoying in compatibility discussion than I have been, and that it might be valuable to have a tool that gathers the effective NUnit API surface used by popular NuGet packages so we know ahead of time who we will be breaking and why.

**Comment by MaceWindu** [comment]
(just to ping author) @zastrowm

**Comment by CharliePoole** [comment]
@MaceWindu @zastrowm `TestContext` is part of the NUnit API for test writers and was intended to be readonly. This change was therefore treated as a bug. Custom applications have other facilities, generally in the Internal namespace that they can use. In this case, if you want to set a property you should use `TestExecutionContext.CurrentContext.CurrentTest.Properties`. Note that `CurrentTest` is NUnit's actual internal representation of the test itself, so you have to tread carefully. The property you set will be reflected to users of your custom attribute in the user-facing `TextContext`. This change should work effectively with all releases of NUnit since 3.0.

@jnm2 I disagree. We knew quite well what we were changing and what it might break. What we didn't know, obviously, is whose code it would break. If we wanted (and had time) to be more aware of that, we could do what I did for NUnit V2: set up a mechanism for extenders to tell us about their extensions, test them and certify them. It could be some sort of contrib repository or simply a place to list them on the website. I can tell you in advance that it's a big job to keep up with extension, however.

**Comment by jnm2** [comment]
> In this case, if you want to set a property you should use `TestExecutionContext.CurrentContext.CurrentTest.Properties`

In my opinion, recommending that is confirming that it is public API. 
We simply don't have time to lead people through the pitfalls of the subtle and unusual difference NUnit maintains between the public visibility mechanism and the blessed API.

**Comment by MaceWindu** [comment]
Can confirm that proposed solution to use `TestExecutionContext` works

**Comment by jnm2** [comment]
In other words, we can give ourselves whatever permissions we want to break people, but I would really rather prioritize not breaking people. It's not as though there is a way for users and extenders to know ahead of time what APIs we will consider buggy or internal.

Can we think about working on 4.0 to end up with an absolutely minimal public API and everything else internal, and do all our cleanup (such as this issue) in there?

**Comment by CharliePoole** [comment]
@jnm2 For the sake of newcomers reading the thread, I'll recap what we have discussed many times before...

1. "Public API" is too broad a concept to be useful to us in many situations.
2. APIs need to be distinguished as to the target audience that __uses__ the API. (This is a concept established years ago, by the way, long before .NET and C# existed.)
3. Since we cannot __prevent__ the wrong audience from using a particular API, we have to document it and warn them against it.
4. Off the top of my head, NUnit supports at least the following APIs:
  * The test writer API (of which `TestContext` forms a part)
  * The framework extension API (of which `TestExecutionContext` forms a part, along with `Test`, `TestResult` and various other things.
  * The framework execution API, used by our own and third-party runners that deal directly with the framework, as well as by test engine drivers.
  * The Test Engine API, used by our own and third-party runners that work through the engine
 * The Test Engine extensibility API, used by those writing engine extensions.

Your work in detecting whether we have broken anything in the public surface of the framework is very useful, but it's far from the whole picture. It's a failing in NUnit that we have not documented the distinctions among the various APIs. Ignoring the distinction will not solve the problem, however.

**Comment by jnm2** [comment]
I like that. Doesn't that mean TestExecutionContext, as framework extension API, needs to come out of the NUnit.Framework.Internal namespace? Its current location is sending conflicting messages.

**Comment by rprouse** [comment]
@jnm2 If we do a massive cleanup and only expose a minimal API, it will likely be a barrier to updating. I think we should continue to try to not break people, but it is going to happen. In this particular case, it ended up being a minor update. I just updated GitHub's OctoKit in another project and it had massive breaking changes in a minor release. In another project a few days ago, the same with a command line parser. I don't want to get that bad, but we can't guarantee all of our APIs are stable. It isn't true SemVer, but it is the best we can do.

That said, going forward, we probably shouldn't be making APIs internal in 3.x releases unless they are dangerous or misused. I also agree with your comments about the internal namespace, it is confusing and sends mixed messages, but it is the reality we need to live with.

**Comment by CharliePoole** [comment]
@jnm2 "In other words, we give ourselves whatever permissions we want to break people..."

This is facile and disingenuous. You are well aware of prior discussions on this topic and of the fact that all of us are very concerned with compatibility. Speaking or writing in over-simplified slogans doesn't help. Presenting yourself as the only team member concerned with such things doesn't help. We have had a lot of past discussions about the proper __strategy__ to avoid breaking users, not about whether it's a good idea to break them and it seems that you are knowingly misrepresenting that past discussion.

As laid out in the earlier post and in prior discussions, I believe that the distinction between "public API" and "everything else" is too simplistic for an application as complex as NUnit. If you only have one API, then the simplified approach can work. But if you have multiple publicly accessible Types, intended for various purposes, you have to use a variety of means to ensure that people know what those purposes are and how changeable each API may be, since they will obviously differ.

**Comment by CharliePoole** [comment]
@rprouse In the past we discussed the possibility of documenting (re-documenting, in fact) multiple APIs versus the public versus non-public approach that @jnm2 advocated. I argued for multiple APIs, pretty much along the lines of my comment above. You (?) or the team decided to go with the public versus non-public approach. IIRC, the reasons given were that it is simpler and that it is a starting point.

If we had taken the approach of identifying and documenting those Types and methods that are used when extending NUnit, we might be in a different place right now. I doubt that we would have finished with the extensibility API, because it's a hard job, but more things would be soe sort of Extensibility namespace rather than internal. Other things might have become protected rather than public or internal. But it all has to start with definition and identification.

I agreed in the past that @jnm2's work on automatically identifying the public, external surface of the framework could be useful as a starting point. However, I haven't seen it used that way so far. I moved on because of lack of interest in my ideas and nobody else has wanted to pick it up. That said, some work on what a custom attribute writer needs to be able to access would be a big step in the right direction, if somebody wanted to take it on.

**Comment by jnm2** [comment]
Ouch. I said "We _can_ give," meaning, it is _possible_ for us to justify almost any change as a bug or internal. Whether something is a feature or a bug is up to potentially varying perspectives. And so much is in the Internal namespace that isn't internal, and out of it which is internal.
In contrast, documentation of the APIs you outlined above looks great and would give us something to point extension writers to.

Where am I presenting myself as the only member of the team who care? My attempt to poke fun at myself by saying I would be extra annoying in future compatibility discussions? (Is it working? lol.) You and I disagree at times, but that in itself is not equivalent to me believing you don't care about compatibility! Why would I believe that?

> This is facile and disingenuous.

Stating that I am being superficial and and insincere detracts from the discussion we should be having and is quite deflating. What is your goal by saying this?

**Comment by CharliePoole** [comment]
I corrected the quotation, which I had typed in manually. The word "can" doesn't seem to change the meaning at all. The discussion of multiple APIs is not new. When we had it before you objected to the approach. The team and the team leader subsequently chose a different approach, which you had championed, that of distinguishing the public surface of the framework from the non-public surface.

Yes I read the statement as your claiming to be the only one who cares about the issue. Your "poking fun" at yourself in a separate comment, to which I didn't reply, isn't very relevant.

"Facile and ingenuous" is harsh, I agree. In many circumstances, I wouldn't say it even though I believe it. But once somebody goes beyond the bounds of civility and makes a personal attack, then I think you have to call them on it. That was my goal, since you ask.

I think that's all I need to say here. I'll follow up offline.

**Comment by OsirisTerje** [comment]
@jnm2 Is this issue open or closed?
